h2quic/server.go

```diff
@@ -238,6 +209,44 @@ func (s *Server) handleRequest(session streamCreator, headerStream quic.Stream,
 	return nil
 }
 
+func serveHTTP(handler http.Handler, responseWriter *responseWriter, req *http.Request, streamEnded bool, reqBody *requestBody) {
+	panicked := false
+	func() {
+		defer func() {
+			if p := recover(); p != nil {
+				// Copied from net/http/server.go
+				const size = 64 << 10
+				buf := make([]byte, size)
+				buf = buf[:runtime.Stack(buf, false)]
+				utils.Errorf("http: panic serving: %v\n%s", p, buf)
+				panicked = true
+			}
+		}()
+		handler.ServeHTTP(responseWriter, req)
+	}()
+	if panicked {
+		responseWriter.WriteHeader(500)
+	} else {
+		responseWriter.WriteHeader(200)
+	}
+	if responseWriter.dataStream != nil {
+		if !streamEnded && !reqBody.requestRead {
+			responseWriter.dataStream.Reset(nil)
+		}
+		responseWriter.dataStream.Close()
+	}
+}
+
+func hasStreamEnded(h2headersFrame *http2.HeadersFrame, dataStream quic.Stream) bool {
```

Comment:
A function called `has...` shouldn't modify state (which you are doing by conditionally closing the stream). I'd rename it to `handleStreamEnded` or similar.